File: 수학_소수/프로그래머스lv1-소수찾기.py
def solution(n):
    # 0 부터 n까지 소수의 배열을 만들어내는 함수
    def makePrimeNumbers(n:int)->list:
        prime_numbers = [True] * (n + 1)
        prime_numbers[0:2] = [False, False]
        for num in range(2, int(n * 0.5) + 1):
            if prime_numbers[num]:
                for i in range(num ** 2, n + 1, num):
                    prime_numbers[i] = False
        return prime_numbers

    prime = makePrimeNumbers(n)

    return prime[:n + 1].count(True)

"""
정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.4MB)
테스트 2 〉	통과 (0.03ms, 10.3MB)
테스트 3 〉	통과 (0.10ms, 10.2MB)
테스트 4 〉	통과 (0.18ms, 10.4MB)
테스트 5 〉	통과 (0.09ms, 10.2MB)
테스트 6 〉	통과 (0.97ms, 10.3MB)
테스트 7 〉	통과 (0.54ms, 10.3MB)
테스트 8 〉	통과 (1.17ms, 10.2MB)
테스트 9 〉	통과 (1.14ms, 10.3MB)
테스트 10 〉	통과 (35.11ms, 14.3MB)
테스트 11 〉	통과 (114.53ms, 23.5MB)
테스트 12 〉	통과 (39.41ms, 14.7MB)
효율성  테스트
테스트 1 〉	통과 (128.41ms, 24.3MB)
테스트 2 〉	통과 (121.79ms, 23.7MB)
테스트 3 〉	통과 (125.39ms, 24.2MB)
테스트 4 〉	통과 (125.85ms, 24.1MB)
"""

File: 수학_소수/프로그래머스lv1-소수 찾기.js
function makePrimeNumbers(n) {
  const primeNumbers = Array.from({ length: n + 1 }, (v) => true)
  primeNumbers.splice(0, 2, false, false)
  for (let num = 2; num < Math.floor(Math.sqrt(n)) + 1; num++) {
    if (primeNumbers[num]) {
      for (let i = num * num; i < n + 1; i += num) {
        primeNumbers[i] = false
      }
    }
  }
  return primeNumbers
}

function solution(n) {
  const primeNumbers = makePrimeNumbers(n)
  return primeNumbers.reduce((acc, num) => (num ? acc + num : acc), 0)
}

// 정확성  테스트
// 테스트 1 〉	통과 (0.25ms, 30.3MB)
// 테스트 2 〉	통과 (0.18ms, 30.1MB)
// 테스트 3 〉	통과 (0.37ms, 30.3MB)
// 테스트 4 〉	통과 (0.30ms, 30.1MB)
// 테스트 5 〉	통과 (0.43ms, 30.1MB)
// 테스트 6 〉	통과 (1.86ms, 30.1MB)
// 테스트 7 〉	통과 (0.73ms, 30.2MB)
// 테스트 8 〉	통과 (1.33ms, 30.3MB)
// 테스트 9 〉	통과 (4.37ms, 30.1MB)
// 테스트 10 〉	통과 (58.00ms, 34.3MB)
// 테스트 11 〉	통과 (181.27ms, 38.9MB)
// 테스트 12 〉	통과 (49.86ms, 34.6MB)
// 효율성  테스트
// 테스트 1 〉	통과 (152.31ms, 39MB)
// 테스트 2 〉	통과 (144.37ms, 38.9MB)
// 테스트 3 〉	통과 (149.04ms, 39.1MB)
// 테스트 4 〉	통과 (139.24ms, 38.9MB)
